lib/bgpstream_bgpdump.c

```diff
@@ -0,0 +1,234 @@
+  char *buf_p = buf;
+
+  /* Record type */
+  switch (elem->type) {
+  case BGPSTREAM_ELEM_TYPE_RIB:
+    c = snprintf(buf_p, B_REMAIN, "TABLE_DUMP2|%u", record->time_sec);
+    break;
+  case BGPSTREAM_ELEM_TYPE_ANNOUNCEMENT:
+  case BGPSTREAM_ELEM_TYPE_WITHDRAWAL:
+  case BGPSTREAM_ELEM_TYPE_PEERSTATE:
+    c = snprintf(buf_p, B_REMAIN, "BGP4MP|%u", record->time_sec);
+    break;
+  default:
+    break;
+  }
+  written += c;
+  buf_p += c;
+  ADD_PIPE;
+
+  switch(elem->type){
+  case BGPSTREAM_ELEM_TYPE_RIB:
+    c = snprintf(buf_p, B_REMAIN, "B");
+    break;
+  case BGPSTREAM_ELEM_TYPE_ANNOUNCEMENT:
+    c = snprintf(buf_p, B_REMAIN, "A");
+    break;
+  case BGPSTREAM_ELEM_TYPE_WITHDRAWAL:
+    c = snprintf(buf_p, B_REMAIN, "W");
+    break;
+  case BGPSTREAM_ELEM_TYPE_PEERSTATE:
+    c = snprintf(buf_p, B_REMAIN, "STATE");
+    break;
+  default:
+    break;
+  }
+
+  written += c;
+  buf_p += c;
+  ADD_PIPE;
+
+  /* PEER IP */
+  if (bgpstream_addr_ntop(buf_p, B_REMAIN, &elem->peer_ip) == NULL) {
+    bgpstream_log(BGPSTREAM_LOG_ERR, "Malformed peer address");
+    return NULL;
+  }
+  SEEK_STR_END;
+  ADD_PIPE;
+
+  /* PEER ASN */
+  c = snprintf(buf_p, B_REMAIN, "%" PRIu32, elem->peer_asn);
+  written += c;
+  buf_p += c;
+  ADD_PIPE;
+
+  switch(elem->type){
+  case BGPSTREAM_ELEM_TYPE_RIB:
+  case BGPSTREAM_ELEM_TYPE_ANNOUNCEMENT:
+    /* PREFIX */
+    if (bgpstream_pfx_snprintf(buf_p, B_REMAIN,
+                               (bgpstream_pfx_t *)&(elem->prefix)) == NULL) {
+      bgpstream_log(BGPSTREAM_LOG_ERR, "Malformed prefix (R/A)");
+      return NULL;
+    }
+    SEEK_STR_END;
+    ADD_PIPE;
+
+    /* AS PATH */
+    c = bgpstream_as_path_snprintf(buf_p, B_REMAIN, elem->as_path);
+    written += c;
+    buf_p += c;
+    if (B_FULL)
+      return NULL;
+    ADD_PIPE;
+
+    /* SOURCE (IGP) */
+    switch(elem->origin){
+    case BGPSTREAM_ELEM_BGP_UPDATE_ORIGIN_IGP:
+      c = snprintf(buf_p, B_REMAIN, "IGP");
+      break;
+    case BGPSTREAM_ELEM_BGP_UPDATE_ORIGIN_EGP:
+      c = snprintf(buf_p, B_REMAIN, "EGP");
+      break;
+    case BGPSTREAM_ELEM_BGP_UPDATE_ORIGIN_INCOMPLETE:
+      c = snprintf(buf_p, B_REMAIN, "INCOMPLETE");
+      break;
+    default:
+      break;
+    }
+    written += c;
+    buf_p += c;
+    ADD_PIPE;
+
+    /* NEXT HOP */
+    if (bgpstream_addr_ntop(buf_p, B_REMAIN, &elem->nexthop) != NULL) {
+      SEEK_STR_END;
+    }
+    ADD_PIPE;
+
+    /* LOCAL_PREF */
+    c = snprintf(buf_p, B_REMAIN, "%d", elem->local_pref);
+    written += c;
+    buf_p += c;
+    ADD_PIPE;
+
+    /* MED */
+    c = snprintf(buf_p, B_REMAIN, "%d", elem->med);
+    written += c;
+    buf_p += c;
+    ADD_PIPE;
+
+    /* COMMUNITIES */
+    c = bgpstream_community_set_snprintf(buf_p, B_REMAIN, elem->communities);
+    written += c;
+    buf_p += c;
+    if (B_FULL)
+      return NULL;
+    ADD_PIPE;
+
+    /* AGGREGATE AG/NAG */
+    if(elem->atomic_aggregate == 1){
```

Comment:
Should we have an enum for these values?

Reply:
it's a boolean value, either set or not set. i don't think enum for these is necessary.